addons/product/models/product_attribute_value.py

```diff
@@ -98,26 +93,35 @@ def write(self, values):
             self.env.invalidate_all()
         return res
 
-    @api.ondelete(at_uninstall=False)
-    def _unlink_except_used_on_product(self):
+    def get_error_message(self):
         for pav in self:
             if pav.is_used_on_products:
-                raise UserError(_(
+                return _(
                     "You cannot delete the value %(value)s because it is used on the following "
-                    "products:\n%(products)s\n If the value has been associated to a product in the"
-                    " past, you will not be able to delete it.",
+                    "products:\n%(products)s\n",
                     value=pav.display_name,
                     products=", ".join(pav.pav_attribute_line_ids.product_tmpl_id.mapped('display_name')),
-                ))
+                )
+        return False
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_except_used_on_product(self):
+        for pav in self:
+            if (error_message := pav.get_error_message()):
+                raise UserError(error_message)
+
+    def unlink(self):
+        for pav in self:
             linked_products = pav.env['product.template.attribute.value'].search(
                 [('product_attribute_value_id', '=', pav.id)]
             ).with_context(active_test=False).ptav_product_variant_ids
             unlinkable_products = linked_products._filter_to_unlink()
-            if linked_products != unlinkable_products:
-                raise UserError(_(
-                    "You cannot delete value %s because it was used in some products.",
-                    pav.display_name
-                ))
+            active_linked_products = linked_products.filtered('active')
+            if linked_products != unlinkable_products or not active_linked_products:
+                # If attributes found on archived product variants
+                # archive PAV instead of deleting
```

Comment:
are you sure it includes all cases because here https://github.com/odoo/odoo/blob/1e6923833113010e14c46dc89ee39617446c97a8/addons/stock/models/product.py#L618-L622 for ex. it checks that the product is used in `stock.lot`. Or maybe I am missing smth?

Reply:
yes, agree